cwms-data-api/src/main/java/cwms/cda/data/dao/RateDao.java

```diff
@@ -0,0 +1,186 @@
+            Timestamp ratingTimstamp = input.getRatingTime().map(Timestamp::from).orElse(null);
+            STR_TAB_T independentTsIds = new STR_TAB_T(input.getTimeSeriesIds());
+            return CWMS_RATING_PACKAGE.call_RETRIEVE_RATED_TS(context.configuration(), independentTsIds, ratingId,
+                input.getOutputUnit(), Timestamp.from(input.getStartTime()), Timestamp.from(input.getEndTime()),
+                ratingTimstamp, "UTC", formatBool(input.getRound()), formatBool(input.getTrim()),
+                formatBool(input.getStartInclusive()), formatBool(input.getEndInclusive()),
+                formatBool(input.getPrevious()), formatBool(input.getNext()), version, "T", officeId,
+                officeId);
+        });
+        List<TimeSeries.Record> records = ztsvTypes.stream()
+            .map(z -> new TimeSeries.Record(z.getDATE_TIME(), z.getVALUE(), z.getQUALITY_CODE().intValue()))
+            .collect(toList());
+        return new RatedOutputTimeSeries(CwmsId.buildCwmsId(officeId, ratingId), records, input.getOutputUnit());
+    }
+
+    public RatedOutput reverseRate(String officeId, String ratingId, RateInputTimeSeries input) {
+        //Performing early simple validation in order to avoid validation within the database
+        validateReverseRateInput(input);
+        ZTSV_ARRAY ztsvTypes = connectionResult(c -> {
+            DSLContext context = getDslContext(c, officeId);
+            Timestamp version = input.getVersionDate().map(Timestamp::from).orElse(null);
+            Timestamp ratingTimstamp = input.getRatingTime().map(Timestamp::from).orElse(null);
+            String timeseriesId = input.getTimeSeriesIds().get(0);
+            return CWMS_RATING_PACKAGE.call_RETRIEVE_REVERSE_RATED_TS(context.configuration(), timeseriesId, ratingId,
+                input.getOutputUnit(), Timestamp.from(input.getStartTime()), Timestamp.from(input.getEndTime()),
+                ratingTimstamp, "UTC", formatBool(input.getRound()), formatBool(input.getTrim()),
+                formatBool(input.getStartInclusive()), formatBool(input.getEndInclusive()),
+                formatBool(input.getPrevious()), formatBool(input.getNext()), version, "T", officeId,
+                officeId);
+        });
+        List<TimeSeries.Record> records = ztsvTypes.stream()
+            .map(z -> new TimeSeries.Record(z.getDATE_TIME(), z.getVALUE(), z.getQUALITY_CODE().intValue()))
+            .collect(toList());
+        return new RatedOutputTimeSeries(CwmsId.buildCwmsId(officeId, ratingId), records, input.getOutputUnit());
+    }
+
+    private <R> R connectionResult(ConnectionCallable<R> callable) {
+        try {
+            return connectionResult(dsl, callable);
+        } catch(DataAccessException ex) {
+            throw handleRateDbError(ex);
+        }
+    }
+
+    static RuntimeException handleRateDbError(DataAccessException ex) {
+        RuntimeException retval = ex;
```

Comment:
See Ryan's message on another PR: https://github.com/USACE/cwms-data-api/pull/1093#issuecomment-2873710523

This may be appropriate as-is, but possible it should be organized with the other JOOQ/Database derived exception handling.

Reply:
I wanted to specifically target the exception codes thrown by the rate functions in the cwms_rating package, I don't necessarily want to globally accept database error messages from other packages. I should add a comment here explaining why it isn't part of the base class exception handling though.

Reply:
Yeah, that makes sense.